lib/endpoints/class-wp-rest-posts-terms-controller.php

```diff
@@ -0,0 +1,225 @@
+			return new WP_Error( 'rest_post_taxonomy_invalid', __( 'Invalid taxonomy for post ID.' ), array( 'status' => 404 ) );
+		}
+
+		$terms = wp_get_object_terms( $post->ID, $this->taxonomy );
+
+		$response = array();
+		foreach ( $terms as $term ) {
+			$data = $this->terms_controller->prepare_item_for_response( $term, $request );
+			$response[] = $this->prepare_response_for_collection( $data );
+		}
+
+		$response = rest_ensure_response( $response );
+
+		return $response;
+	}
+
+	/**
+	 * Get a term that is attached to a post
+	 *
+	 * @param WP_REST_Request $request Full details about the request
+	 * @return WP_Error|WP_REST_Response
+	 */
+	public function get_item( $request ) {
+		$post     = get_post( $request['id'] );
+		$term_id  = absint( $request['term_id'] );
+
+		$post_check = $this->posts_controller->get_item( $request );
+		if ( is_wp_error( $post_check ) ) {
+			return $post_check;
+		}
+
+		$valid_taxonomies = get_object_taxonomies( $post->post_type );
+		if ( ! in_array( $this->taxonomy, $valid_taxonomies ) ) {
+			return new WP_Error( 'rest_post_taxonomy_invalid', __( 'Invalid taxonomy for post ID.' ), array( 'status' => 404 ) );
+		}
+
+		if ( ! get_term_by( 'term_taxonomy_id', $term_id, $this->taxonomy ) ) {
+			return new WP_Error( 'rest_term_invalid', __( "Term doesn't exist." ), array( 'status' => 404 ) );
+		}
+
+		$terms = wp_get_object_terms( $post->ID, $this->taxonomy );
+
+		if ( ! in_array( $term_id, wp_list_pluck( $terms, 'term_taxonomy_id' ) ) ) {
+			return new WP_Error( 'rest_post_not_in_term', __( 'Invalid taxonomy for post ID.' ), array( 'status' => 404 ) );
+		}
+
+		$term = $this->terms_controller->prepare_item_for_response( get_term_by( 'term_taxonomy_id', $term_id, $this->taxonomy ), $request );
+
+		$response = rest_ensure_response( $term );
+
+		return $response;
+	}
+
+	/**
+	 * Add a term to a post
+	 *
+	 * @param WP_REST_Request $request Full details about the request
+	 * @return WP_Error|WP_REST_Response
+	 */
+	public function create_item( $request ) {
+		$post     = get_post( $request['id'] );
+		$term_id  = absint( $request['term_id'] );
+
+		$post_check = $this->posts_controller->get_item( $request );
+		if ( is_wp_error( $post_check ) ) {
+			return $post_check;
+		}
+
+		$valid_taxonomies = get_object_taxonomies( $post->post_type );
+		if ( ! in_array( $this->taxonomy, $valid_taxonomies ) ) {
+			return new WP_Error( 'rest_post_taxonomy_invalid', __( 'Invalid taxonomy for post ID.' ), array( 'status' => 404 ) );
+		}
+
+		if ( ! get_term_by( 'term_taxonomy_id', $term_id, $this->taxonomy ) ) {
+			return new WP_Error( 'rest_term_invalid', __( "Term doesn't exist." ), array( 'status' => 404 ) );
+		}
+
+		$tt_ids = wp_set_object_terms( $post->ID, $term_id, $this->taxonomy, true );
+
+		if ( is_wp_error( $tt_ids ) ) {
+			return $tt_ids;
+		}
+
+		$term = $this->terms_controller->prepare_item_for_response( get_term_by( 'term_taxonomy_id', $term_id, $this->taxonomy ), $request );
```

Comment:
We already have the tight coupling to the `terms_controller` here in several other places, so I think this should remain consistent with using that. For example we use `get_item` in the validation method. Also, if a route is removed for `/terms/post_tag` for whatever reason, internally routing will not return what is expected.